src/main/kotlin/com/openlattice/indexing/BackgroundLinkingIndexingService.kt

```diff
@@ -103,54 +113,92 @@ class BackgroundLinkingIndexingService(
     private val unIndexCandidates = hazelcastInstance
             .getQueue<Triple<List<Array<UUID>>, UUID, OffsetDateTime>>(HazelcastQueue.LINKING_UNINDEXING.name)
 
+    private val limiter = Semaphore(indexerConfiguration.parallelism)
+
     @Suppress("UNUSED")
-    private val linkingIndexingWorker = executor.submit {
-        generateSequence(indexCandidates::take)
-                .chunked(LINKING_INDEX_SIZE)
-                .asStream()
-                .parallel()
-                .forEach { candidateBatch ->
-                    if (!indexerConfiguration.backgroundLinkingIndexingEnabled) {
-                        return@forEach
-                    }
-
-                    val (linkingEntityKeyIdsWithLastWrite, linkingIds) = mapCandidates(candidateBatch)
-
-                    try {
-                        lock(linkingIds)
-                        index(linkingEntityKeyIdsWithLastWrite, linkingIds)
-                    } catch (ex: Exception) {
-                        logger.error("Unable to index linking entity with from batch of linking ids $linkingIds.", ex)
-                    } finally {
-                        unLock(linkingIds)
-                    }
-                }
+    @Scheduled(fixedRate = LINKING_INDEX_RATE)
+    private fun linkingIndexing() {
+        submitIndexingTask(indexCandidates, true)
+    }
+
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = LINKING_INDEX_RATE)
+    private fun linkingUnIndexing() {
+        submitIndexingTask(unIndexCandidates, false)
+    }
+
+    private fun submitIndexingTask(
+            candidates: IQueue<Triple<List<Array<UUID>>, UUID, OffsetDateTime>>,
+            createMode: Boolean
+    ) {
+        if (!isLinkingIndexingEnabled()) {
+            return
+        }
+
+        val taskName = if (createMode) "index" else "un-index"
+        executor.submit {
+            try {
+                logger.info("Starting background linking $taskName task.")
+                val esw = Stopwatch.createStarted()
+
+                generateSequence(candidates::poll)
+                        .chunked(LINKING_INDEX_SIZE)
+                        .forEach { candidateBatch ->
+
+                            limiter.acquire()
```

Comment:
as a novice at threads sorry if I'm wrong, but shouldn't this happen before/right after the submit call?

Reply:
In this case, we use the limiter `Semaphore` to limit the number of jobs submitted to a certain number and not for locking the whole process